test_contextual_lenses/test_learning.py

```diff
@@ -0,0 +1,454 @@
+"""End-to-end learning tests for contextual_lenses.py."""
+
+
+import jax.numpy as jnp
+
+import numpy as np
+
+from absl.testing import parameterized
+from absl.testing import absltest
+
+from contextual_lenses import mean_pool, max_pool, \
+linear_max_pool, linear_mean_pool, gated_conv 
+
+from train_utils import create_optimizer, train_step, \
+create_representation_model
+
+from encoders import one_hot_encoder, cnn_one_hot_encoder, \
+one_hot_pos_emb_encoder, cnn_one_hot_pos_emb_encoder
+
+
+def generate_random_sequences(bs=3, seq_len=12, num_categories=21):
+  """Generates bs many random sequences of length seq_len from num_categories indices."""
+
+  np.random.seed(0)
+  input_data = jnp.array(np.random.randint(0, num_categories-1, size=(bs, seq_len)))
+  
+  np.random.seed(0)
+  output_data = jnp.array(np.random.normal(size=(bs)))
+
+  return input_data, output_data
+
+
+def compute_train_loss(model, input_data, output_data, learning_rate=1e-3, epochs=100):
```

Comment:
I'd call this train(). To me, compute_train_loss() would compute the training loss wrt the current model parameters. Here, you're actually optimizing doing training.